File: ruby/dnd-character/dnd_character.rb
class DndCharacter

  BASE_HITPOINTS = 10
  private_constant :BASE_HITPOINTS

  private
  def initialize
    @strength     = roll_die
    @dexterity    = roll_die
    @constitution = roll_die
    @intelligence = roll_die
    @wisdom       = roll_die
    @charisma     = roll_die
    @hitpoints    = BASE_HITPOINTS + self.class.modifier(constitution)
  end
  
  private
  def self.modifier(ability_score)
    ability_score / 2 - 5 
  end

  public
  def roll_die(roll: 4, die: 6, remove_lowest: 1)
    roll.times.map { rand(1..die) }.sort.pop(roll - remove_lowest).sum
  end

  public
  attr_reader :strength,
    :dexterity,
    :constitution,
    :intelligence,
    :wisdom,
    :charisma,
    :hitpoints
end

File: typescript/dnd-character/_iterations/1/dnd-character.ts
export class DnDCharacter {
  readonly strength: number;
  readonly dexterity: number;
  readonly constitution: number;
  readonly intelligence: number;
  readonly wisdom: number;
  readonly charisma: number;
  readonly hitpoints: number;

  constructor() {
    this.strength = DnDCharacter.generateAbilityScore()
    this.dexterity = DnDCharacter.generateAbilityScore()
    this.constitution = DnDCharacter.generateAbilityScore()
    this.intelligence = DnDCharacter.generateAbilityScore()
    this.wisdom = DnDCharacter.generateAbilityScore()
    this.charisma = DnDCharacter.generateAbilityScore()
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution)
  }
  
  public static generateAbilityScore(): number {
    const rolls: number[] = [...Array(4)].map(_ => Math.floor(Math.random() * 6) + 1)
    return rolls.sort((a, b) => a - b).slice(0, 3).reduce((n, sum) => sum + n)
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2)
  }
}
